src/darwin/Framework/CHIP/MTRLogging.mm

```diff
@@ -0,0 +1,70 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTRLogging_Internal.h"
+
+#import "MTRFramework.h"
+
+#import <algorithm>
+#import <atomic>
+#import <os/lock.h>
+
+using namespace chip::Logging;
+
+static_assert(MTRLogTypeError == (NSInteger) kLogCategory_Error, "MTRLogType* != kLogCategory_*");
+static_assert(MTRLogTypeProgress == (NSInteger) kLogCategory_Progress, "MTRLogType* != kLogCategory_*");
+static_assert(MTRLogTypeDetail == (NSInteger) kLogCategory_Detail, "MTRLogType* != kLogCategory_*");
+
+static os_unfair_lock logCallbackLock = OS_UNFAIR_LOCK_INIT;
+static MTRLogCallback logCallback = nil;
+
+static void MTRLogCallbackTrampoline(const char * component, uint8_t category, const char * format, va_list args)
```

Comment:
I expect that if you annotate this function with `ENFORCE_FORMAT(3, 0)` you would not need to do manual -Wformat-nonliteral suppression, because then this function would assume that the enforcement happens in the caller (which it would, if it were not being called via function pointer).

Reply:
It doesn't quite work without warnings because we're going from a format in an `char *` here to an `NSString` with `@(format)`, so even with `ENFORCE_FORMAT` the compiler warns about that expression not being a literal when passed to `initWithFormat:`. I think we'll have to live with the pragmas at this point.